enterprise/server/remote_execution/containers/firecracker/firecracker.go

```diff
@@ -945,15 +945,36 @@ func (c *FirecrackerContainer) saveSnapshot(ctx context.Context, snapshotDetails
 
 	vmd := c.getVMMetadata().CloneVT()
 	vmd.LastExecutedTask = c.getVMTask()
+
+	// Only save a remote snapshot if a remote snapshot for the primary git branch key
+	// doesn't already exist. Otherwise, only save a local snapshot. Workloads
+	// should hit an executor with a local snapshot due to affinity routing on
+	// the git branch key.
+	//
+	// We want to always save the master snapshot remotely, because it is used as
+	// a fallback for runs on other branches, and we want the latest master snapshot available
+	// to all executors. `!c.hasFallbackKeys()` is a good proxy for whether we're
+	// running on the master snapshot.
+	hasRemoteSnapshotForBranchKey := c.hasRemoteSnapshotForBranchKey(ctx)
+	shouldCacheRemotely := c.supportsRemoteSnapshots &&
+		(!c.hasFallbackKeys() || !hasRemoteSnapshotForBranchKey)
+	if !shouldCacheRemotely && c.supportsRemoteSnapshots {
+		log.CtxInfof(ctx, "Would not save remote snapshot. "+
+			"Has remote snapshot for branch key: %v. Snapshot keys: %v", hasRemoteSnapshotForBranchKey, c.snapshotKeySet)
+	}
+	if shouldCacheRemotely && hasRemoteSnapshotForBranchKey {
+		log.CtxInfof(ctx, "Would save remote snapshot even though one already exists. Snapshot keys: %v", c.snapshotKeySet)
+	}
 	opts := &snaploader.CacheSnapshotOptions{
-		VMMetadata:          vmd,
-		VMConfiguration:     c.vmConfig,
-		VMStateSnapshotPath: filepath.Join(c.getChroot(), snapshotDetails.vmStateSnapshotName),
-		KernelImagePath:     c.executorConfig.GuestKernelImagePath,
-		InitrdImagePath:     c.executorConfig.InitrdImagePath,
-		ChunkedFiles:        map[string]*copy_on_write.COWStore{},
-		Recycled:            c.recycled,
-		Remote:              c.supportsRemoteSnapshots,
+		VMMetadata:                 vmd,
+		VMConfiguration:            c.vmConfig,
+		VMStateSnapshotPath:        filepath.Join(c.getChroot(), snapshotDetails.vmStateSnapshotName),
+		KernelImagePath:            c.executorConfig.GuestKernelImagePath,
+		InitrdImagePath:            c.executorConfig.InitrdImagePath,
+		ChunkedFiles:               map[string]*copy_on_write.COWStore{},
+		Recycled:                   c.recycled,
+		Remote:                     shouldCacheRemotely,
```

Comment:
this needs to still be set to `c.supportsRemoteSnapshots`, right?

Reply:
Oops yes - good catch